### DISCUSSION TITLE
sql injection prevention methods

### DISCUSSION THREAD

**Discussion by 3l3ktr0n** [e107inc/e107#4813] [General]
Does e107 have any integrated measures for preventing sql injection?

**Comment by 3l3ktr0n** [comment]
hmm no answer eh?

**Comment by Deltik** [comment]
@3l3ktr0n: e107 does have [some support for prepared statements](https://github.com/e107inc/e107/blob/31d7002ff258738b39b780ae7d50e72d21b96bc5/e107_handlers/e_db_pdo_class.php#L354-L380); however, the interface could be improved.

For legacy reasons, most client code that builds SQL queries uses [`\e_parse::toDB()`](https://github.com/e107inc/e107/blob/31d7002ff258738b39b780ae7d50e72d21b96bc5/e107_handlers/e_parse_class.php#L488-L637) to escape untrusted input.

It is my wish to move to only prepared statements in the future, but part of the challenge is that the output format of `\e_parse::toDB()` is loaded with a lot of baggage, so cleanly separating these usages in client code will require a lot of educated guesses and database migrations, which could cause data loss if done improperly.

**Reply by 3l3ktr0n** [reply]
Thank you for the answer. As far as I understood toDB(); escapes only quotes and doubles what aren't even needed for sql injection

**Reply by Deltik** [reply]
Right, that's what I meant by "`\e_parse::toDB()` is loaded with a lot of baggage".  It escapes quotes as if they were HTML, even for data types that aren't HTML.

There is currently no official guidance on how to reconcile this complication, but my personal view is to use prepared statements where possible.  `\e_db_pdo::db_Query()` has [documentation](https://github.com/e107inc/e107/blob/31d7002ff258738b39b780ae7d50e72d21b96bc5/e107_handlers/e_db_pdo_class.php#L313-L328) on e107's limited support for prepared statements.

**Reply by 3l3ktr0n** [reply]
Well I ain't the right person to tell as everyones grandma codes better than me.. but in my opinion this legacy commitment is holding back e107s full potential. You can't cater to everyone. It's like trying to be the best ay everything. You can be the best at one thing or weak at everything. I'd ditch the chatbox and replace it with a messenger (something facebook like), instead of just display all the comments and forums posts maybe create a user timeline, add somekind of group functionality maybe? It's nice to get notifications in your mailbox but it'd be even nicer to get them on the site as seperate notification system or via pm. Some of the social networking themes are kind of the norm these days. But that's my opinion. If this where to change I think e107 would blow other cms out of the water. It feels as it is still in the 2000's.
This isn't to bring anyone down. Don't get me wrong I LOVE e107 BUT DAMN IS THIS ONE HARD RELATIONSHIP. My to cents if anyone cares. Maybe even work in the direction of being a framework rather than a cms?

**Reply by Deltik** [reply]
I know what you mean.  It's hard to attract good developers due to how ancient the codebase is.  I am a developer with lofty visions for e107, but I have too many time commitments elsewhere to be able to do the overhauls necessary to modernize the CMS.

**Reply by 3l3ktr0n** [reply]
is it me or only pdo has some kind of support for prepared statements but mysql has none?

**Reply by Deltik** [reply]
That's right.  `e_db_mysql` currently does not support prepared statements, but `e_db_pdo` does.  I don't believe there is a reason to use `e_db_mysql` at all, though.

**Comment by 3l3ktr0n** [comment]
`$sql->select('user', 'user_id, user_name', 'user_id=:id OR user_name=:name ORDER BY user_name', array('id' => 999, 'name'=>'e107'));`

has anyone tried binding like this from the manual?

**Reply by Deltik** [reply]
The e107 automated tests have [sample usages of parameter binding](https://github.com/e107inc/e107/blob/bd9f62675bfb250e5406dee789de391c69c539ea/e107_tests/tests/unit/e_db_pdoTest.php#L82-L137).

[`e_db_pdoTest::testSelectBind()`](https://github.com/e107inc/e107/blob/bd9f62675bfb250e5406dee789de391c69c539ea/e107_tests/tests/unit/e_db_pdoTest.php#L86) demonstrates the exact query you mentioned, and [`e_db_pdoTest::testDb_QueryBind()`](https://github.com/e107inc/e107/blob/bd9f62675bfb250e5406dee789de391c69c539ea/e107_tests/tests/unit/e_db_pdoTest.php#L96) is an example of an `INSERT INTO` query.

**Reply by 3l3ktr0n** [reply]
hmm for some reason it is not working for me

**Reply by Deltik** [reply]
What have you tried?  We need a code sample and some context to be able to troubleshoot your issue.

**Reply by 3l3ktr0n** [reply]
I tried the sample bind code that's in the testSelectBind in my file. I get SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '= :game_id AND review_id = :review_id' at line 1
```
    $query = array(
        'PREPARE' => 'SELECT * FROM ' . MPREFIX . 'game_comment WHERE game_id=:game_id AND review_id=:review_id',
        'EXECUTE' => array(
            ':game_id' => $gameID,
            ':review_id' => $reviewID
        )
    );
    $result = $sql->db_Query($query, null, 'db_Select');
```

**Reply by Deltik** [reply]
The bound arguments should be specified without the leading colon (`:`).

**Reply by 3l3ktr0n** [reply]
nothing did change

**Reply by 3l3ktr0n** [reply]
my bad, commented out some code and 'it works' now.. seems other queries where interfering.. now I have to get it to print

well... or not

**Reply by 3l3ktr0n** [reply]
```
$sql->select('game_comment', 'game_id, user_id, review_id', 'game_id=:game AND review_id=:review AND user_id=:user', ['game' => 1, 'review' => 1, 'user' => 1], true);
print_a($sql);
while($result = $sql->fetch())
{
    print_a($result);
}
```
gives
```
[mySQLaccess:protected] => PDO Object
        (
        )

    [mySQLresult] => 
    [mySQLrows] => Array
        (
            [0] => 0
        )

    [mySQLerror] => 1
    [mySQLlastErrNum:protected] => 42000
    [mySQLlastErrText:protected] => SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '=:game AND review_id=:review AND user_id=:user' at line 1
    [mySQLlastQuery:protected] => SELECT game_id, user_id, review_id FROM e107_game_comment game_id=:game AND review_id=:review AND user_id=:user
    [mySQLcurTable] => game_comment
    [mySQLlanguage] => 
    [mySQLinfo] => 
    [tabset] => 1
```
and
```
$query = array(
			'PREPARE' => 'SELECT * FROM ' . MPREFIX . 'game_comment WHERE game_id=:game_id AND user_id=:user_id',
			'EXECUTE' => array(
				'game_id'   => 1,
				'user_id'   => 1
			)
		);
		$sql->db_Query($query, null, 'db_Select');
        print_a($sql);
        while($res = $sql->fetch())
        {
            echo $res['game_id'];
        }
```
gives
```
[mySQLaccess:protected] => PDO Object
        (
        )

    [mySQLresult] => 
    [mySQLrows] => Array
        (
            [0] => 0
        )

    [mySQLerror] => 1
    [mySQLlastErrNum:protected] => 0
    [mySQLlastErrText:protected] => 
    [mySQLlastQuery:protected] => Array
        (
            [PREPARE] => SELECT * FROM e107_game_comment WHERE game_id=:game_id AND user_id=:user_id
            [EXECUTE] => Array
                (
                    [game_id] => 1
                    [user_id] => 1
                )

        )
```
nothing to fetch

**Reply by 3l3ktr0n** [reply]
am I missing something or?

**Reply by Deltik** [reply]
Are you using e107 v2.2.0 or newer?  The output you showed looked like it came from the old `e_db_mysql` class, which shouldn't even be possible to use without [modding the core](https://github.com/e107inc/e107/blob/6b828a9ac5bfa645955998e902757b9a7ea14d4d/e107_handlers/e107_class.php#L170-L171).  `e_db_pdo::select()` got the ability to bind parameters in starting in e107 v2.2.0.

**Reply by Jimmi08** [reply]
Isn't here missing WHERE? ` SELECT game_id, user_id, review_id FROM e107_game_comment game_id=:game AND review_id=:review AND user_id=:user`

**Reply by Deltik** [reply]
Yes, the object was of type the old `e_db_mysql`, and they called 

<pre>public function e_db_mysql::select(
     $table,
     $fields = '*',
     $arg = '',
     <strong>$noWhere</strong> = false,
     $debug = FALSE,
     $log_type = '',
     $log_remark = '' ): false|int|string</pre>

with `$noWhere` set to a truthy value, which removes the `WHERE`.

**Reply by 3l3ktr0n** [reply]
oh man, I was on 2.2.0 but for the db it was still mysql_class. Might have been an error when updating while back. Now on 2.3.2 and all is good. Thanks
